src/components/OasisMarketWidget.js

```diff
@@ -1,16 +1,82 @@
-import React, { PureComponent } from 'react'
+import React, { PureComponent } from 'react';
 import { PropTypes } from 'prop-types';
+import { last } from 'lodash';
+
 import OasisWidgetFrame from '../containers/OasisWidgetFrame';
+import { OasisTable } from './OasisTable';
+import { ETH_UNIT_ETHER } from '../constants';
+import { TIME_SPAN_DAY, TIME_SPAN_MONTH, TIME_SPAN_WEEK } from '../utils/timeSpan';
+import getTokenPairPrice from '../utils/tokens/getTokenPairPrice';
+import getTokenPairVolume from '../utils/tokens/getTokenPairVolume';
+import getTokenPairTrades from '../utils/tokens/getTokenPairTrades';
+import getTokenPairFormat from '../utils/tokens/getTokenPairFormat';
+
+const TOKEN_PAIR_MARKET_VOLUME_DECIMALS = 2;
+const TOKEN_PAIR_MARKET_PRICE_DECIMALS = 4;
+
+const timeSpanHeading = {
+  [TIME_SPAN_DAY]: 'daily',
+  [TIME_SPAN_WEEK]: 'weekly',
+  [TIME_SPAN_MONTH]: 'monthly',
+};
 
+const getMarketColDefinition = (timeSpan) => {
+  return [
+    { heading: 'pairs', key: 'tokenPair' },
+    { heading: `${timeSpanHeading[timeSpan]} volume`, key: 'volume' },
+    { heading: 'price', key: 'tradingPairPrice' },
+  ];
+};
+
+const formatTokenPairPrice =
+  (price) =>
+    price ? price.toFormat(TOKEN_PAIR_MARKET_PRICE_DECIMALS) : null;
 
 class OasisMarketWidget extends PureComponent {
-  render(){
+  constructor(props) {
+    super(props);
+    this.transformRow = this.transformRow.bind(this);
+  }
+
+  transformRow(row) {
+    const { marketData = [] } = this.props;
+    const [baseToken, quoteToken] = [row.get('base'), row.get('quote')];
+    const tokenPairFormat = getTokenPairFormat(baseToken, quoteToken);
+
+    if (marketData) {
+      const tokenPairTrades = getTokenPairTrades(marketData, baseToken, quoteToken);
+      const tradingPairVolume = getTokenPairVolume(tokenPairTrades, baseToken);
+      const tradingPairPrice = tradingPairVolume.toNumber() ?
+        getTokenPairPrice(last(tokenPairTrades), baseToken, quoteToken) : null;
+
+      return {
+        tokenPair: getTokenPairFormat(baseToken, quoteToken),
+        volume: window.web3.fromWei(tradingPairVolume, ETH_UNIT_ETHER).toFormat(TOKEN_PAIR_MARKET_VOLUME_DECIMALS),
+        tradingPairPrice: formatTokenPairPrice(tradingPairPrice),
+      };
+    } else {
+      return {
```

Comment:
One method that returns different interfaces doesn't seem right. That assumes that either I have to check the returned value  if it has given properties everytime I call this method or there will be potencial null pointers

Reply:
Yes You have to check if this a null each time.

Reply:
I'm not talkinga bout null , I'm talking about the different object structure

Reply:
They look same to me

Reply:
oh, You are right.

Reply:
done